### DISCUSSION TITLE
Scala 3.7 given priority change causing ambiguous givens

### DISCUSSION THREAD

**Discussion by yoohaemin** [scalalandio/chimney#592] [General]
Not an issue right now, but probably going to be in the near future...

https://www.scala-lang.org/2024/08/19/given-priority-change-3.7.html

I've turned on -source 3.6 and now I'm seeing these warnings:

```
-- Warning: /path/to/source.scala:39:33
39 |        _.transformInto[Bar]
   |                       ^
   |Given search preference for io.scalaland.chimney.Transformer.AutoDerived[Foo, Bar] between alternatives
   |  (Foo.given_Transformer_Foo_Bar : io.scalaland.chimney.Transformer[Foo, Bar])
   |and
   |  (io.scalaland.chimney.Transformer.AutoDerived.deriveAutomatic : [From, To]: io.scalaland.chimney.Transformer.AutoDerived[From, To])
   |will change.
   |Current choice           : the first alternative
   |New choice from Scala 3.7: none - it's ambiguous
```

Right now I have a transformer instance defined (unnamed given) in the companion object of `Foo` (different source) and I have dsl.* imported at the callsite.

**Comment by MateuszKubuszok** [comment]
I'll contact the compiler devs about it. For now I am not touching this - it could be solved with e.g. replacing each

```scala
final def transformInto[To](implicit transformer: Transformer.AutoDerived[From, To]): To =
  transformer.transform(source)
```

with

```scala
inline def transformInto[To]: To = summonFrom {
  case transformer: Transformer[From, To] => transformer.transform(source)
  case _                                  => Transformer.AutoDerived.derive
}
```

or similar, but that would break binary compatibility, bumping version to 2.0.0... I'm going to delay that until I get more info from the compiler team.

In the meantime - these warnings and priorities should only be for `given`s, not for `implicit`s - if you rename your anonymous `given` to a named implicit it should disappear.

**Reply by yoohaemin** [reply]
Thanks for the reply. I am safe until 3.7 is released, unless I decide to use the new priorities before 3.7 release (unlikely). This is more of a heads-up :smile: 

There are plenty of workarounds, so no worries.

**Reply by MateuszKubuszok** [reply]
Yeah, there are, although I'd prefer to:

 - focus on some remaining long pending issues in the next following months
 - make sure that all "planned" features are there
 - before I'll even consider some cleanup that would justify breaking bincompat and releasing 2.0.0
 - so that people on 1.x would get as much as possible, hassle of moving to 2.x should be justified.

Ideally, I'd want out users to get the same source compatibility even after the major version bump, especially since we are trying to keep the code compiling in 2.12/2.13/3.

This change can make it tricky sometimes, e.g. when someone would use:

```scala
implicit def keyConverter[KeyA, KeyB](
  // allows using user provider Transformer or fallback on generated Transformer.AutoDerived
  implicit transformer: Transformer.AutoDerived[From, To]
): KeyConverter[KeyA, KeyB] = ...
```

then whatever we would use in core library would not automatically propagate to our users and that could badly reflect on us :/.

**Reply by MateuszKubuszok** [reply]
I tried to reproduce the issue and fix it. From what I see you have it when you store the `given`s in your companion objects. Unfortunately, the [implicit priorities](https://www.scala-lang.org/2024/08/19/given-priority-change-3.7.html#explicit-prioritization-by-owner) do [**not** work](https://github.com/scalalandio/chimney/pull/617). I [pinned down the current behavior](https://github.com/scalalandio/chimney/pull/618) so that future Community Build runs would show the compiler team that the issue exists (IMHO the fact that so little projects had issues is a testament of not testing things that should obviously work rather than negligible impact of that change).

**Reply by MateuszKubuszok** [reply]
I found 1 way of working around the issue - https://scastie.scala-lang.org/eKrwKqN8S2uHp5pjr61uiw or https://scastie.scala-lang.org/KXeEJJxHTWSpZIM640FM6A - but I am not sure whether it going to be source-compatible change, while it very likely to be a binary incompatible change - something that would require releasing Chimney 2.0.0 much faster than I'd wanted to.

**Comment by MateuszKubuszok** [comment]
How we'll solve it depends on https://github.com/scala/scala3/discussions/21909 - if it's merged, it will let us simplify Chimney 2.0.0 for next S3 LTS. If not, it will be a PITA.

Currently, I am certain that Chimney 1.+ will never support semiautomatic derivation with new `given`s.

**Comment by MateuszKubuszok** [comment]
I tested locally that for 3.7.0-RC1 you can work around and use `foo.into[Bar].transform` and it doesn't cause any issues where `transformInto` fails.

I guess sometime soon I'll create 2.0.0-M1 with [`summonIgnore`](https://github.com/scala/scala3/discussions/21909)

**Reply by MateuszKubuszok** [reply]
WIP: https://github.com/scalalandio/chimney/pull/713

**Comment by MateuszKubuszok** [comment]
Released support for `3.7.0` changes in [`2.0.0-M1`](https://github.com/scalalandio/chimney/releases/tag/2.0.0-M1) - once we have next Scala 3 LTS, and we would be sure that nothing more needs to be changed, we will release it as stable `2.0.0`.
